File: server-side/src/api/product/product.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Product } from './entities/product.entity';
import { CreateProductDto } from './dto/create-product.dto';
import { UpdateProductDto } from './dto/update-product.dto';

@Injectable()
export class ProductService {
  constructor(
    @InjectRepository(Product)
    private readonly productRepo: Repository<Product>,
  ) {}

  async getAllProducts(): Promise<Product[]> {
    return this.productRepo.find();
  }

  async getProductById(id: number): Promise<Product> {
    const product = await this.productRepo.findOneBy({ id });
    if (!product) {
      throw new NotFoundException(`Product with ID ${id} not found`);
    }
    return product;
  }

  async createProduct(data: CreateProductDto): Promise<Product> {
    const product = this.productRepo.create(data);
    return this.productRepo.save(product);
  }

  async updateProduct(id: number, data: UpdateProductDto): Promise<Product> {
    const existingProduct = await this.getProductById(id);
    await this.productRepo.update(id, data);
    return this.getProductById(id);
  }

  async deleteProduct(id: number): Promise<boolean> {
    const existingProduct = await this.getProductById(id);
    const result = await this.productRepo.delete(id);
    return (result.affected ?? 0) > 0;
  }
}

File: backend-server/src/main/java/DM/com/backend_server/service/ProductService.java
package DM.com.backend_server.service;


import DM.com.backend_server.Entity.Product;
import DM.com.backend_server.dto.CreateProductDto;
import DM.com.backend_server.dto.UpdateProductDto;
import DM.com.backend_server.exception.ProductNotFoundException;
import DM.com.backend_server.repository.ProductRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class ProductService {

    private final ProductRepository productRepository;

    @Autowired
    public ProductService(ProductRepository productRepository) {
        this.productRepository = productRepository;
    }

    public List<Product> getAllProducts() {
        return productRepository.findAll();
    }

    public Product getProductById(Long id) {
        return productRepository.findById(id)
                .orElseThrow(() -> new ProductNotFoundException("Product with ID " + id + " not found"));
    }

    public Product createProduct(CreateProductDto data) {
        Product product = new Product();
        // Map DTO fields to entity (manually or use MapStruct/ModelMapper for complex cases)
        product.setName(data.getName());
        product.setDescription(data.getDescription());
        product.setPrice(data.getPrice());
        // Set other fields
        return productRepository.save(product);
    }

    public Product updateProduct(Long id, UpdateProductDto data) {
        Product existingProduct = getProductById(id);
        // Update fields (partial update)
        if (data.getName() != null) {
            existingProduct.setName(data.getName());
        }

        if (data.getPrice() != null) {
            existingProduct.setPrice(data.getPrice());
        }
        // Update other fields as needed
        return productRepository.save(existingProduct);
    }

    public boolean deleteProduct(Long id) {
        Product existingProduct = getProductById(id);
        productRepository.deleteById(id);
        return true; // Since deleteById doesn't return affected count easily, assume success after existence check
    }
}
